src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/ViewProviders/GraphViewProviderBase.cs

```diff
@@ -126,29 +126,29 @@ public virtual bool ShouldTrackChanges(string projectPath, string updatedProject
             ITargetedDependenciesSnapshot targetedSnapshot,
             IDependency updatedDependency,
             GraphNode dependencyGraphNode,
-            out IEnumerable<DependencyNodeInfo> nodesToAdd,
-            out IEnumerable<DependencyNodeInfo> nodesToRemove)
+            out IReadOnlyList<DependencyNodeInfo> nodesToAdd,
+            out IReadOnlyList<DependencyNodeInfo> nodesToRemove)
         {
-            IEnumerable<DependencyNodeInfo> existingChildrenInfo = GetExistingChildren(dependencyGraphNode);
+            IReadOnlyList<DependencyNodeInfo> existingChildrenInfo = GetExistingChildren(dependencyGraphNode);
             IEnumerable<IDependency> updatedChildren = targetedSnapshot.GetDependencyChildren(updatedDependency)
                 ?? Enumerable.Empty<IDependency>();
-            IEnumerable<DependencyNodeInfo> updatedChildrenInfo = updatedChildren.Select(x => DependencyNodeInfo.FromDependency(x));
+            IReadOnlyList<DependencyNodeInfo> updatedChildrenInfo = updatedChildren.Select(x => DependencyNodeInfo.FromDependency(x)).ToList();
```

Comment:
There could also be an argument for removing the `.ToList()` in the `AnyChanges` method and here - could prevent any construction of DependencyNodeInfo if the circumstances are right. I'm just thinking out loud though mainly.

Reply:
The `Any` calls would kick off an enumeration to the first item. Doing so creates an `Enumerable.ExceptIterator` which fully consumes one enumerable to create a set to compare against. `TrackChanges` fully enumerates both results too. Without `ToList` we end up allocating redundant sets internally to support `Except`.